### ISSUE TITLE
KRt_from_projection returns a reflection (det R = -1) for -P and for P scaled below eps; OpenCV returns a rotation

### ISSUE THREAD

**Issue by ducha-aiki** [kornia/kornia#4864]
`KRt_from_projection` factors `P = K [R | t]` with an RQ decomposition and then flips signs so that `K` has a positive diagonal, using `signs = torch.sign(torch.diagonal(upper_mat) + eps)` with `eps = 1e-6`. Two kinds of valid input come back with a reflection (`det R = -1`) where the docstring promises a rotation matrix. (a) `-P` is the same camera, since `P` is defined up to a nonzero scale, sign included (Hartley–Zisserman §6.1), but it returns `det R = -1` and `t` negated. (b) Because `eps` is added before the sign is taken, a raw diagonal entry in `(-eps, 0)` is not flipped: `1e-7 * P` returns `K[2, 2] = -1e-7`, `det R = -1` and `t[2]` negated. In every case `K [R | t]` still reproduces the input, so the result is a valid factorisation and nothing signals a problem, but `R` is not a rotation. OpenCV's `decomposeProjectionMatrix` returns `det R = +1` for all five scales below.

## Repro

```python
import torch
from kornia.geometry.conversions import axis_angle_to_rotation_matrix
from kornia.geometry.epipolar import KRt_from_projection, projection_from_KRt

K = torch.tensor([[[700.0, 0.0, 300.0], [0.0, 740.0, 240.0], [0.0, 0.0, 1.0]]])
R = axis_angle_to_rotation_matrix(torch.tensor([[0.1, -0.2, 0.05]]))
t = torch.tensor([[[0.5], [0.1], [0.02]]])
P = projection_from_KRt(K, R, t)

for s in [1.0, 2.0, -1.0, 1e-7, -1e-7]:
    Kd, Rd, td = KRt_from_projection(s * P)
    rel = ((projection_from_KRt(Kd, Rd, td) - s * P).abs().max() / (s * P).abs().max()).item()
    print(f"s={s:>5g}: diag K {[round(v, 9) for v in torch.diagonal(Kd[0]).tolist()]}, "
          f"det R {torch.linalg.det(Rd[0]).item():+.4f}, t {[round(v, 4) for v in td[0, :, 0].tolist()]}, "
          f"|K[R|t] - sP| / |sP| {rel:.0e}")
```

```
s=    1: diag K [699.999938965, 740.0, 0.99999994], det R +1.0000, t [0.5, 0.1, 0.02], |K[R|t] - sP| / |sP| 1e-07
s=    2: diag K [1399.99987793, 1480.0, 1.999999881], det R +1.0000, t [0.5, 0.1, 0.02], |K[R|t] - sP| / |sP| 1e-07
s=   -1: diag K [699.999938965, 740.0, 0.99999994], det R -1.0000, t [-0.5, -0.1, -0.02], |K[R|t] - sP| / |sP| 1e-07
s=1e-07: diag K [7e-05, 7.4e-05, -1e-07], det R -1.0000, t [0.5, 0.1, -0.02], |K[R|t] - sP| / |sP| 1e-07
s=-1e-07: diag K [7e-05, 7.4e-05, 1e-07], det R -1.0000, t [-0.5, -0.1, -0.02], |K[R|t] - sP| / |sP| 1e-07
```

The `s = 1e-7` row is case (b). The `s = -1e-7` row is the control: its raw entry is `+1e-7`, it is not flipped, and only case (a) applies.

OpenCV on the same matrices, run separately with `uv run --no-project --with opencv-python-headless --with numpy python`:

```python
import cv2
import numpy as np

K = np.array([[700.0, 0.0, 300.0], [0.0, 740.0, 240.0], [0.0, 0.0, 1.0]])
R = cv2.Rodrigues(np.array([0.1, -0.2, 0.05]))[0]
t = np.array([[0.5], [0.1], [0.02]])
P = K @ np.hstack([R, t])

for s in [1.0, 2.0, -1.0, 1e-7, -1e-7]:
    Kc, Rc, C = cv2.decomposeProjectionMatrix(s * P)[:3]
    tc = -Rc @ (C[:3] / C[3])  # OpenCV returns the camera centre C; t = -R C
    print(f"s={s:>5g}: diag K {[round(v, 9) for v in np.diag(Kc).tolist()]}, "
          f"det R {np.linalg.det(Rc):+.4f}, t {[round(v, 4) for v in tc[:, 0].tolist()]}")
```

```
s=    1: diag K [700.0, 740.0, 1.0], det R +1.0000, t [0.5, 0.1, 0.02]
s=    2: diag K [1400.0, 1480.0, 2.0], det R +1.0000, t [0.5, 0.1, 0.02]
s=   -1: diag K [700.0, 740.0, -1.0], det R +1.0000, t [-0.5, -0.1, 0.02]
s=1e-07: diag K [7e-05, 7.4e-05, 1e-07], det R +1.0000, t [0.5, 0.1, 0.02]
s=-1e-07: diag K [7e-05, 7.4e-05, -1e-07], det R +1.0000, t [-0.5, -0.1, 0.02]
```

OpenCV keeps `det R = +1` in every row. For a negative scale it lets `K[2, 2]` go negative instead, so it does not recover the original `(R, t)` for `-P` either. It has no `eps` shift, so `1e-7 * P` decomposes exactly.

kornia main 7ddf731bd, torch 2.14.0 (and 2.5.1: same), CPU float32, macOS arm64; OpenCV 5.0.0.

## Classification

Silent wrong result. Case (b) is Idiom 1 (`x + eps` perturbing a result instead of guarding a singularity): `eps` moves the sign test, so it changes the output rather than guarding against anything. Case (a) is Idiom 4 (degenerate or singular input silently producing garbage instead of raising): a `P` whose left `3 x 3` block has negative determinant is a valid camera, and it silently gets a non-rotation.

## Disposition

Focused fix, welcome as a PR: first multiply `P` by `sign(det(P[:, :, :3]))`, the same sign ambiguity that HZ's depth formula (§6.2.3) resolves with `sign(det M)`. Then replace `torch.sign(diag + eps)` with `torch.where(diag < 0, -1, 1)`, which never produces a zero sign. Checked out of tree by wrapping the current function (sign-normalised `P`, `eps=0.0`) in float64: `s = ±1`, `±1e-7` and `1e-30` all return `det R = +1`, the positive-diagonal `K` scaled by `|s|`, and `R`, `t` within `2.8e-16` of the originals. Inputs with `det > 0` and no raw diagonal entry in `(-eps, 0)` give unchanged results, because the two sign rules agree on them. Two documented behaviours change: `-P` returns the original rotation and translation instead of a reflection and `-t`, and `eps` no longer affects the result.
Found by the batch-8 conventions audit (two-view geometry); a `test_wart_*` pin in the docs PR flips when the fix lands.

Posted on behalf of @ducha-aiki by Claude (Fable 5.1).
